# Simple script showing how to pull data from Expression Atlas and recount3, normalize, and remove batch effects.

```{r setup, include=FALSE}
library(ExpressionAtlas)
library(plyr)
library(tidyverse)
library(limma)
library(sva)
library(edgeR)
library(ggplot2)
library(janitor)
library(foreach)
library(doParallel)
library(biomaRt)
library(ggfortify)
library(patchwork)
library(cowplot)
library(clusterProfiler)
library(AnnotationDbi)
library(org.Hs.eg.db)
pak::pkg_install("rrcov")
library(rrcov)

source("functions.R")

# Set timeout to avoid failure when trying to download GTEx or other large datasets
options(timeout = 1800)
```

```{r}
experimental_metadata <- read_csv("expression_atlas_metadata.csv")
experimental_metadata <- experimental_metadata[-1,] # Remove the top row (GTEx) to lower the time cost when testing

# Load metadata from recount3
# experimental_metadata <- read_csv("recount3_metadata.csv")

plots_dir <- "data/"

# Setup filters for removing samples accordingly to the metadata
feature_vec <- list()
feature_vec[["disease"]] <- c("normal", "control", "", NA, 
                              "non inflammatory bowel disease control")
```


## recount3
```{r recount3}
# This is commented out as we're sticking with Expression Atlas for now.

# library(recount3)

# human_projects <- available_projects()

# proj_info <- subset(
#   human_projects,
#   project == "SRP166108" & project_type == "data_sources"
# )
# rse <- create_rse(proj_info)

# #
# map_to_cols <- function(s) {
#   s %>%
#     str_split("\\|", simplify = TRUE) %>%
#     str_split(";;", simplify = TRUE) %>%
#     as_tibble() %>%
#     t() %>%
#     row_to_names(row_number = 1) %>%
#     repair_names()
# }

# convert_metadata_to_df <- function(sample_attributes) {
#   meta <- lapply(sample_attributes, FUN = map_to_cols)
#   meta <- dplyr::bind_rows(meta)
#   meta <- DataFrame(meta)
#   rownames(meta) <- rownames(colData(rse))
#   return(meta)
# metadata_df <- convert_metadata_to_df(colData(rse)$sra.sample_attributes)
# rse@colData <- metadata_df
# }
```

## Expression Atlas

```{r}
accession_ids <- experimental_metadata$id

rnaseq_exps <- getAtlasData(accession_ids)

all_exps <- rnaseq_exps
exps <- names(rnaseq_exps@listData)

exp_data <- lapply(exps, FUN = function(x) {
    return(all_exps[[x]]$rnaseq)
})
names(exp_data) <- exps
```



## Main loop
```{r}
source("./main_processing_loop.R")

parallel = FALSE
if(.Platform$OS.type == "unix") {
  parallel = TRUE
  library(doMC)
  registerDoMC(32)
} 
results_list <- llply(names(exp_data), main_count_processing, exp_data, experimental_metadata, .parallel = parallel)  
names(results_list) <- names(exp_data)
for(dset_name in names(results_list)) 
  save_plot(filename=paste0(plots_dir, dset_name,"_pca.png"), results_list[[dset_name]]$plotPanel, 
            base_height = 6, base_asp = 1.2, ncol = 2, nrow = 2)

save(results_list, file = "results_list.RData")
```

Now we calculate the gene expression variance in all the samples

```{r}

for (dset_name in names(results_list)) {
  print(dset_name)
  exprDf = results_list[[dset_name]]
  gene_vars = rowVars(exprDf)
  names(gene_vars) = 
  }
```

